### ISSUE TITLE
Idea: Add per-environment global variables

### ISSUE THREAD

**Issue by kydorn** [markround/tiller#18]
It would add a lot of flexibility being able to specify global environment variables, that apply only to a specific environment, but to all templates. 

Like any other global variable, it could be overriden by template-specific variables.

**Comment by markround** [comment]
Hi, Thanks for the suggestion. Do you have an example for how you'd like to see it working ?

**Comment by kydorn** [comment]
Sure, for example:
`
data_sources: [ 'defaults', 'file' ]
template_sources: [ 'file' ]

defaults:
    global:
      variable: 'default value'
    template.erb:
       ...
    template2.erb:
       ...
environments:
    development:
      global: 
        variable: 'default value for development environment'
      template.erb
         variable: 'local value for template.erb'
    staging:
      global: 
        variable: 'default value for staging environment'
    production:
     ...
`
This is specially useful when we are sharing the same variable between multiple templates and different environments. If we need the same variable to be consistent in an environment, we currently have to define it for each template that uses it

**Comment by markround** [comment]
Right, I see. Let me have a think about this, and I'll see if I can add something like that for you.

**Comment by markround** [comment]
OK, had a think about it. It seems that the best approach would be to make the defaults plugin aware of any settings inside a `common:` block inside the environment, as that's how you over-ride other settings from the 'root' of common.yaml. I have a proof-of-concept that looks like this :

```(yaml)
exec: ['date']
data_sources: [ 'defaults','file','environment' ]
template_sources: [ 'file' ]

defaults:
  global:
    some_global_value: 'This is available to all environments'
    per_env: 'This is the default for the per-env global'

environments:

  development:
    common:
      defaults:
        global:
          per_env: 'This is a per-env global'

    test.erb:
      target: /tmp/test.txt
      config:
        some_local_value: 'This is the development environment'

    test2.erb:
      target: /tmp/test2.txt
      config:
        some_local_value: 'This is the development environment'
        per_env: 'This is the per-env global overwritten by the local value'

  production:

    test.erb:
      target: /tmp/test.txt
      config:
        some_local_value: 'This is the production environment'
```

So, the production environment gets `per_env` set to 'This is the default for the per-env global'.

The development environment has 2 templates, the first uses the `per_env` value from that environment, which is set to 'This is a per-env global', and the second over-rides it with a local value, so it becomes 'This is the per-env global overwritten by the local value'.

Does that do what you want ?

**Comment by markround** [comment]
Actually, forget that. It won't work with v1 configurations. Having another crack at it now...

**Comment by markround** [comment]
Just committed b24073d253abfef0fe75d965692500fd67fd3add , which I think will do what you're after. Take a look at the test fixtures:

 * [v2 format](https://github.com/markround/tiller/tree/feature/per_env_global/features/fixtures/file_globals/v2)
 * [v1 format](https://github.com/markround/tiller/tree/feature/per_env_global/features/fixtures/file_globals/v1)

It looks pretty much like what you were after. It was a very simple change, I just made the file datasource pick up a new `global_values:` block, and you can also use the defaults datasource to specify a default.

Is this OK for you ?

**Comment by kydorn** [comment]
Yup, that would do the trick! Thank you so much for this fast response :)

**Comment by markround** [comment]
Great :) OK, I'll add some notes to the documentation, tidy it up a bit and get a release out ASAP (probably tomorrow). I'll give due credit for the suggestion in the Changelog, would you like me to reference your GitHub ID/website/blog/whatever ?

**Comment by kydorn** [comment]
You could use the handler if you want, but there is no need. I'm actually using Tiller at work for creating a hierarchical configuration for docker containers and I just was too lazy to learn ruby and change it myself. Thank you so much for your work!

**Comment by markround** [comment]
Resolved by [v0.7.7](https://github.com/markround/tiller/tree/0.7.7). An updated gem will be built and released shortly. Thanks again for the suggestion!
